clients/roleProviders/ethereumRoleProvider.go

```diff
@@ -0,0 +1,127 @@
+package roleProviders
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const ethSignatureSize = 64
+
+// ArgsEthereumRoleProvider is the argument for the ethereum role provider constructor
+type ArgsEthereumRoleProvider struct {
+	EthereumChainInteractor EthereumChainInteractor
+	Log                     logger.Logger
+}
+
+type ethereumRoleProvider struct {
+	ethereumChainInteractor EthereumChainInteractor
+	log                     logger.Logger
+	whitelistedAddresses    map[common.Address]struct{}
+	mut                     sync.RWMutex
+}
+
+// NewEthereumRoleProvider creates a new ethereum role provider instance able to fetch the
+// whitelisted addresses and able to check ethereum signatures
+func NewEthereumRoleProvider(args ArgsEthereumRoleProvider) (*ethereumRoleProvider, error) {
+	err := checkEthereumRoleProviderSpecificArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	erp := &ethereumRoleProvider{
+		whitelistedAddresses:    make(map[common.Address]struct{}),
```

Comment:
just a question for myself: should we keep the order of the parameters here same as in the struct?

Reply:
order is not important, for now, in this component.
Reason: this role provider is used to determine if a certain signature received from the p2p subsystem is valid or not, ethereum-wise. That means it does several checks like: public key recovery, signature verification and public key authorization - which is a search in a map fed by data from the ethereum contract

Reply:
understood. thanks